### PR TITLE
Preferred database version

### PR DESCRIPTION
**Pull request by ghiculescu**
A common issue when using Rails in a team is different team members using slightly different database versions, resulting in different `schema.rb` or `structure.sql` files when they run `rake db:migrate`.

This issue was discussed here: https://discuss.rubyonrails.org/t/schema-rb-generation/74640/3

This PR extracts part of an [internal rake task](https://gist.github.com/ghiculescu/234da1a8aada56d543e6b876a17859ff#file-migrate_enhancements-rake-L5..L60) we use to ensure everyone is using the same version of `pg_dump`.

By default there's no changes. If you set `config.active_record.preferred_database_version`, then `rake db:schema:dump` will raise if you're using a version different to the expected one.

(Theoretically we could also use this config var to do things like add warnings in development/test if using the wrong DB version, but for now I'm only tackling the problem of edit wars over schema files.)

### CODE REVIEW

**Review by tenderlove — CHANGES_REQUESTED** [review]
Is there a way to get the database version from the connection rather than shelling out to the program?  The `sqlite3` binary might not be in the user's path, or even worse the version of sqlite3 that was linked against is different than the version being executed (this could be true for all the databases, not just the sqlite3 example)

**Comment by ghiculescu** [comment]
For sqlite3 and mysql, that's probably doable.

For postgres I'll have to dig into it more, since I *think* the structure of `structure.sql` is dependent on the `pg_dump` version, not the actual database version.

File: activerecord/lib/active_record/tasks/sqlite_database_tasks.rb [hunk 1]
```diff
@@ -9,9 +9,8 @@ def self.using_database_configurations?
         true
       end
 
-      def initialize(db_config, root = ActiveRecord::Tasks::DatabaseTasks.root)
+      def initialize(db_config)
         @db_config = db_config
-        @root = root
```
Comment (ghiculescu) [inline activerecord/lib/active_record/tasks/sqlite_database_tasks.rb:14]: not really related to this PR, but this var isn't used for anything so i removed it.

**Comment by ghiculescu** [comment]
@tenderlove I started going down that path, but decided not to because in all 3 cases the command used to dump the schema is the same as the command used to check its version. See [here](https://github.com/rails/rails/blob/master/activerecord/lib/active_record/tasks/sqlite_database_tasks.rb#L57), [here](https://github.com/rails/rails/blob/master/activerecord/lib/active_record/tasks/mysql_database_tasks.rb#L58), and [here](https://github.com/rails/rails/blob/master/activerecord/lib/active_record/tasks/postgresql_database_tasks.rb#L76).

In all 3 cases, Rails already uses `run_cmd` which outputs a helpful message (see `run_cmd_error`) if the command fails, eg. because it's not in path.

To make things simple, I am just not running my check at all if the version command returns nil (ie. it fails). That way I rely on the existing error handling rather than needing to add new error handling.

**Pull request closed without merging**